_posts/2017-08-21-Write-Efficient-Deep-Learning-Operators-with-TVM-A-Depthwise-Convolution-Example.md

```diff
@@ -0,0 +1,412 @@
+---
+layout: post
+title:  "Write Efficient Deep Learning Operators with TVM: A Depthwise Convolution Example"
+date:   2017-08-21
+author: Yuwei Hu
+---
+
+This blog aims to share some guidelines and tricks of using TVM to write efficient deep learning operators.
+It is a summary of what I have learned from the experience of writing the depthwise convolution operator, i.e.
+[topi.nn.depthwise_conv2d_nchw](http://docs.tvmlang.org/api/python/topi.html#topi.nn.depthwise_conv2d_nchw).
+
+For the design philosophy of depthwise convolution, you may refer to [the Xception paper](https://arxiv.org/abs/1610.02357).
+This blog is about its efficient implementation, and basic knowledge of GPU programming (CUDA) is required.
+
+## 0. Introduction
+
+This part briefly talks about three concepts we should know when optimizing CUDA code: data reuse, shared memory, coalesced memory access.
+It would be great if you already know them, then you may skip this part and go directly to part 1.
+
+### Data Reuse
+Data reuse means loading the data once and using it multiple times. Because the time cost of data loading is much higher that computing [1], we always want to reuse data as much as possible.
```

Comment:
Remove the first sentence. 

In modern computing architectures, the cost of loading data from memory is much higher than doing a single floating point computations. Because of that, we always want to reuse the input data after they are being loaded into registers or shared memory(cache).